google/cloud/spanner/internal/time.h

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_INTERNAL_TIME_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_INTERNAL_TIME_H_
+
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <google/protobuf/duration.pb.h>
+#include <google/protobuf/timestamp.pb.h>
+#include <chrono>
+#include <cstdint>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+/**
+ * Convert a std::chrono::nanoseconds to a google::protobuf::Duration.
+ */
+google::protobuf::Duration ToProto(std::chrono::nanoseconds ns);
+
+/**
+ * Convert a google::protobuf::Duration to a std::chrono::nanoseconds.
+ */
+std::chrono::nanoseconds FromProto(google::protobuf::Duration const& proto);
+
+/**
+ * Convert a system_clock::time_point to a google::protobuf::Timestamp.
+ */
+google::protobuf::Timestamp ToProto(std::chrono::system_clock::time_point tp);
+
+/**
+ * Convert a google::protobuf::Timestamp to a system_clock::time_point.
+ */
+std::chrono::system_clock::time_point FromProto(
```

Comment:
There are some platforms where the system_clock duration is not nanoseconds. IIRC, Mac OS X has microsecond resolution on the system clock. Does it matter for these functions if we lose precision when going from protos to system_clock::time_point? 

As an alternative, we could define a nanosecond-resolution time_point on the system_clock, like:

```cc
using time_point_ns = std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
```

... but then the range would be lessened. Bummer.

Another thought would be to require the caller to specify the time-point's duration (resolution), and make all of our functions that take or produce a time_point a template that uses `std::chrono::time_point<std::chrono::system_clock, D>`

Thoughts?

Reply:
My thoughts were that `system_clock::time_point` would be sufficient, but I'm happy to revisit.  I concur that the other alternatives are a particular time_point of sufficient resolution and portable width, or rep/period templates all the way down.  I filed #138 to track this.

[Aside: Our spanci-fedora-libcxx-msan-30 image also has a microsecond system_clock.  The current test cases assume at least that resolution.]